### ISSUE TITLE
Problems with manual reproduction of the Confusion Matrix

### ISSUE THREAD

**Issue by albertalvin8080** [ultralytics/ultralytics#18527]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

*Task*: detect
*n classes*: 1 (egg)
*model*: Yolov10s and Yolov11s

Hello, I'm trying to fine tune Yolov10 and Yolov11 small on some insect eggs images with a really, really noisy background. The only problem is that the confusion matrix ploted by model.val() and the confusion matrix I make manually using model.predict() have different values for TP, FP and FN, and consequently, Precision and Recall values are different as well.

Observations:
- I'm using the same NMS threshold for both (0.7).
- I'm using the same confidence value for both (0.15).
- I'm using the last version of Ultralytics.
- Yes, I'm always using the same _weights_ and _model_ when comparing both ways (ex: yolov11s-1000-epochs-sgd.pt), checked it more than once.

Repository containing both ways of calculating the matrix: https://github.com/albertalvin8080/yolo-metrics-manual
- [model.val() python file](https://github.com/albertalvin8080/yolo-metrics-manual/blob/main/evaluate_metrics_auto_YOLO.py)
- [model.predict() python file](https://github.com/albertalvin8080/yolo-metrics-manual/blob/main/evaluate_metrics_manual_YOLO.py)
- [script used for calculating IOU](https://github.com/albertalvin8080/yolo-metrics-manual/blob/main/some_utils.py)

**Real Example**
For yolov11s, 1000 epochs with SGD, model.val() predicts:
- TP: 90
- FP: 5
- FN: 8

While the code I've written using model.predict() predicts:
- TP: 85
- FP: 18
- FN: 13

The number of TP + FN (i.e. the number of ground truth boxes) is the same for both (98), so i think there's something I'm off about how I should correctly calculate these metrics manually, or something about the internal behavior of model.val() which I don't understand.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @albertalvin8080, thank you for your detailed question and for providing such a clear description with links to your repository 🚀! We appreciate your interest in Ultralytics and efforts to dig deep into the YOLO evaluation metrics.

To better assist you, here are a few suggestions and resources for diagnosing the differences between the confusion matrix generated by `model.val()` and your manual calculations using `model.predict()`:

If this is a 🐛 Bug Report (e.g., unexpected behavior in `model.val()` or discrepancies beyond acceptable numerical tolerances), please provide a full [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) (MRE). This will help us investigate the issue more accurately.

If this is a ❓ Custom Training/Validation Question about implementing your manual metrics, please review the following:
- Ensure your confusion matrix calculation is aligned with the [Ultralytics Metrics implementation](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/utils/metrics.py). Small differences in IoU thresholding or Non-Maximum Suppression (NMS) handling could lead to variations in results.
- Double-check the thresholds (e.g., `NMS=0.7`, `conf=0.15`) being passed to both methods (`model.val()` and your manual calculation).
- Investigate any possible pre-processing or filtering discrepancies between `model.val()` and predictions from `model.predict()`.

## Upgrade

To ensure up-to-date functionality and bug fixes:
Upgrade the `ultralytics` package to the latest version with the minimal required dependencies:

```bash
pip install -U ultralytics
```

For supported environments, verify that you are using one of the pre-configured environments described below, with dependencies like [Python>=3.8](https://www.python.org/) and [PyTorch>=1.8](https://pytorch.org/) properly installed.

## Environment Options

YOLO models are extensively tested across multiple platforms:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud services** with pre-built configurations: See our [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/) and [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) guides.
- **Dockerized Environment**: Use our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) for reproducibility and streamlined setups.

## Cross-Verify Results

Try leveraging the built-in validation available in `model.val()` by running the workflow directly on your dataset with:

```python
from ultralytics import YOLO
model = YOLO('yolov11s.pt')  # replace with your model weight
metrics = model.val()
print(metrics)  # examine built-in metrics
```

If discrepancies still persist, there could be subtle differences in how predictions are processed. Refer to the relevant source code in your shared repository and examine the [val.py logic](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/engine/val.py).

## Community Resources

Join the Ultralytics community for collaborative discussions or immediate troubleshooting:
- Real-time chat via [Discord](https://discord.com/invite/ultralytics) 🎧
- Deep-dive into community questions on our [Discourse](https://community.ultralytics.com) forums or [Subreddit](https://reddit.com/r/Ultralytics) 🧠.

## CI Status

Lastly, check the CI status badge below to ensure the current Ultralytics tests are passing:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

🤖 *This is an automated response. An Ultralytics engineer will review your issue and provide further guidance shortly.*

**Comment by Y-T-G** [comment]
Validation uses 0.25 as confidence threshold

Validation also uses a larger image size due to `rect=True`.

You can try running validation with `rect=False` and `batch=1`.

**Comment by albertalvin8080** [comment]
@Y-T-G Wow, changing rect and batch really made some difference. I tried changing other parameters and found a configuration which comes really close to the results of model.val(). 

Also, instead of passing the image loaded by OpenCV to model.predict(), I'm now passing source=img_path, which increased the number of TPs.

Params passed to model.val():
- `batch=1`
- `rect=False`
- `half=False`
- `conf=0.25`
- `iou=0.7`

Changes to manual prediction:
- `source=img_path` to model.predict()
- `rect=False` to model.predict()
- `half=False` to model.predict()
- `conf=0.25` to model.predict()
- `iou=0.20` to model.predict()
- changed **iou_gt_threshold** to 0.25 (used to manually match bbxs)

Now the confusion matrix looks like this:
- model.predict(): TP=89, FN=9, FP=14
- model.val(): TP=89, FN=9, FP=13

There are some interesting differences for the iou being passed for both, and the values of P and R are still different (most probably the formulas I used differ from the ones used inside mode.val), but it seems good enough for what I'm trying to do.

Thank You!
